### PR TITLE
New mode selector

### PR DESCRIPTION
**Pull request by daniel-heppner-ibigroup**
Implements the new mode selector and GraphQL plan queries.

Will require a weblate update afterwards, en-US merged properly but the rest didn't

### CODE REVIEW

**Comment by daniel-heppner-ibigroup** [comment]
@miles-grant-ibigroup discovered a very strange crash when accessing [this](http://localhost:9966/#/?ui_activeSearch=1yyhdmszs&ui_activeItinerary=-1&fromPlace=33.78325%2C+-84.39406%3A%3A33.783249048627454%2C-84.39406179237416&toPlace=33.8956%2C+-84.09812%3A%3A33.89560378231087%2C-84.09811757850667&date=2023-02-17&time=17%3A14&arriveBy=false&mode=WALK%2CBUS%2CSUBWAY%2CTRAM%2CFLEX_EGRESS%2CFLEX_ACCESS%2CFLEX_DIRECT&showIntermediateStops=true&walkSpeed=1.34&ignoreRealtimeUpdates=true&wheelchair=false&numItineraries=3&otherThanPreferredRoutesPenalty=900&modeButtons=transit_walk) and clicking on any button. I wasn't able to repeat this with any other url, and I am not sure what it is about this specific instance that causes a crash. I fixed it by adding a null check in itineraries header component.

**Comment by miles-grant-ibigroup** [comment]
Another bug: the accessible routing switch doesn't seem to be working properly!

File: lib/components/form/batch-settings.tsx
Comment (binh-dam-ibigroup) [inline lib/components/form/batch-settings.tsx:102]: Can these be moved to `mapStateToProps`? I think redux applies an memo or some optimization that prevents too many frequent changes.

File: lib/components/form/batch-settings.tsx
Comment (binh-dam-ibigroup) [inline lib/components/form/batch-settings.tsx:144]: These methods seem superfluous and should be removed. Use the setter from `useState` directly.

File: lib/components/form/batch-settings.tsx
Comment (binh-dam-ibigroup) [inline lib/components/form/batch-settings.tsx:148]: Delete comment.

File: lib/components/form/batch-settings.tsx
Comment (binh-dam-ibigroup) [inline lib/components/form/batch-settings.tsx:191]: Remove. This no longer applies.

File: lib/components/form/batch-preferences.tsx
Comment (binh-dam-ibigroup) [inline lib/components/form/batch-preferences.tsx:23]: This component can be removed now, do you think?

**Review by binh-dam-ibigroup — CHANGES_REQUESTED** [review]
First round of comments. Mostly cleanup until the mode selector API solidifies.

File: lib/components/form/batch-settings.tsx
Comment (daniel-heppner-ibigroup) [reply lib/components/form/batch-settings.tsx:93]: This variable is keeping track of both the settings element and the datetime element, which is why it's a string.

File: lib/components/form/batch-settings.tsx
Comment (daniel-heppner-ibigroup) [reply lib/components/form/batch-settings.tsx:191]: oh, of course...

File: lib/components/form/batch-settings.tsx
Comment (daniel-heppner-ibigroup) [reply lib/components/form/batch-settings.tsx:102]: I don't think we can do that because of the use of ModeIcon from context, which is only available in a component. I believe I looked into putting this in the `mapStateToProps` when I wrote it.

**Comment by binh-dam-ibigroup** [comment]
I think the main thing missing now is that the mode selector selections aren't really synced with the URL params. One example is that when I set `wheelchair=true` in the URL, the accessible routing box doesn't come out checked as it should.

File: lib/actions/apiV2.js
Comment (binh-dam-ibigroup) [inline lib/actions/apiV2.js:813]: I'm wondering whether the `modeButtons` query param is needed. It mostly mirrors the `mode` query param, right?

**Comment by binh-dam-ibigroup** [comment]
> moved away from Redux-URL State sync

I think being to paste a link and get the mode selected according to that in the UI is an awesome feature.

File: lib/actions/apiV2.js
Comment (daniel-heppner-ibigroup) [reply lib/actions/apiV2.js:813]: I'd like to remove the `mode` queryParam since that is for the old system. Mode buttons do not necessarily map one to one with modes either since a button could activate multiple modes.

**Comment by daniel-heppner-ibigroup** [comment]
> I think the main thing missing now is that the mode selector selections aren't really synced with the URL params. One example is that when I set `wheelchair=true` in the URL, the accessible routing box doesn't come out checked as it should.

It is synced, but the mode settings are all grouped together under `modeSettings` query param. There are not specific query parameters for individual settings, since they're all customizable from the config. If we want to add some hard coded mappings for backwards compatibility, we could check those on application startup for backwards compatibility, but I think that would be quite messy.

File: lib/actions/api.js
Comment (binh-dam-ibigroup) [inline lib/actions/api.js:220]: Remove query parameters from unactivated modes to shorten and clean up the URL. Remove query parameters that are set on their default value.

Comment (binh-dam-ibigroup) [reply lib/actions/api.js:220]: Can we use `:` as separator? We use `::` as separator for query from/to locations.

**Review by binh-dam-ibigroup — CHANGES_REQUESTED** [review]
Remove the unnecesary file changes, and implement removal of mode settings variables for modes that aren't active and variables set to their default value.

File: lib/components/form/plan-trip-button.tsx
Comment (binh-dam-ibigroup) [inline lib/components/form/plan-trip-button.tsx:33]: A11Y: Do not disable the plan trip button, instead, let users click it, and if something is missing, tell them at that point. See https://css-tricks.com/making-disabled-buttons-more-inclusive/, https://www.smashingmagazine.com/2021/08/frustrating-design-patterns-disabled-buttons/

**Review by binh-dam-ibigroup — CHANGES_REQUESTED** [review]
There is one more big blocking issue regarding the mode buttons resetting when clicking the "Plan Trip" button.

File: lib/components/util/styledIcon.tsx
Comment (daniel-heppner-ibigroup) [reply lib/components/util/styledIcon.tsx:57]: bad merge

File: lib/components/form/batch-settings.tsx
Comment (daniel-heppner-ibigroup) [reply lib/components/form/batch-settings.tsx:122]: Are you asking me to make a change besides merging dev?

Comment (binh-dam-ibigroup) [reply lib/components/form/batch-settings.tsx:122]: Merging with dev should be enough.

File: lib/actions/api.js
Comment (daniel-heppner-ibigroup) [reply lib/actions/api.js:217]: This was actually brought up a while ago, I think. The problem with this is that `modes` are different from `modeButtons`, since `modeButtons` refers to the selected buttons, each of which contains various modes in it that get activated or deactivated with the button. 

I've had to re-add support for directly specifying the modes in the currentQuery object for call taker and field trip.

File: lib/components/form/batch-settings.tsx
Comment (daniel-heppner-ibigroup) [reply lib/components/form/batch-settings.tsx:91]: The fix for this involves breaking the saved searches functionality. I think this is something we should chat about, I need another mind to think about a solution.

File: lib/components/form/batch-settings.tsx
Comment (binh-dam-ibigroup) [inline lib/components/form/batch-settings.tsx:58]: Add a brief comment about this function.

File: lib/components/form/batch-settings.tsx
Comment (binh-dam-ibigroup) [reply lib/components/form/batch-settings.tsx:80]: Still needs to be addressed.

File: lib/components/form/batch-settings.tsx
Comment (binh-dam-ibigroup) [inline lib/components/form/batch-settings.tsx:179]:
```suggestion
          onSettingsUpdate={setUrlSearch}
```

File: lib/components/form/batch-styled.ts
Comment (binh-dam-ibigroup) [inline lib/components/form/batch-styled.ts:116]: Can you use the bootstrap thresholds? Otherwise, add a comment to explain the choice of thresholds.

File: lib/components/form/batch-styled.ts
Comment (binh-dam-ibigroup) [inline lib/components/form/batch-styled.ts:117]: This produces the behavior in the screenshot below. Is that desired?

image

File: lib/components/form/batch-styled.ts
Comment (binh-dam-ibigroup) [inline lib/components/form/batch-styled.ts:119]: Sort CSS.

File: lib/components/util/styledIcon.tsx
Comment (binh-dam-ibigroup) [inline lib/components/util/styledIcon.tsx:59]: Remove trailing whitespace.

**Review by binh-dam-ibigroup — COMMENTED** [review]
Reminder about some items still to be resolved, and a few other things.

File: lib/actions/api.js
Comment (binh-dam-ibigroup) [reply lib/actions/api.js:220]: > Remove query parameters from unactivated modes to shorten and clean up the URL. Remove query parameters that are set on their default value.

I think this still needs to be addressed.

File: lib/components/form/batch-settings.tsx
Comment (binh-dam-ibigroup) [reply lib/components/form/batch-settings.tsx:121]: This has been fixed by https://github.com/opentripplanner/otp-react-redux/pull/700/commits/5d881fd3f9b467c153455695d831c502126a4030. Nice job.

File: lib/actions/apiV2.js
Comment (miles-grant-ibigroup) [inline lib/actions/apiV2.js:74]: This might need some work...

Comment (daniel-heppner-ibigroup) [reply lib/actions/apiV2.js:74]: This code doesn't look familiar to me at all

Comment (binh-dam-ibigroup) [reply lib/actions/apiV2.js:74]: What does this have to do with the mode selector?

Comment (daniel-heppner-ibigroup) [reply lib/actions/apiV2.js:74]: Actually, since the functions are no longer used anywhere else, I just deleted them from the api.js file. I think at this point any further refactoring can be left to the future OTPv1 removal PR, which will change all of this.

**Comment by miles-grant-ibigroup** [comment]
Percy is still flagging 2 issues: accessibility scores are still missing and trip data isn't coming back either

File: lib/actions/apiV2.js
Comment (binh-dam-ibigroup) [inline lib/actions/apiV2.js:876]: Some params aren't being parsed correctly. For instance, removing the bus option still results in itineraries with a bus leg being returned (see screenshot).

image

It looks like boolean params are converted to strings in the OTP-UI package in use, which probably resolve to `true` (see debugger screenshot).
image

**Review by binh-dam-ibigroup — CHANGES_REQUESTED** [review]
Setting state for checkboxes seems to be parsed incorrectly and needs to be fixed. See also other minor items.

File: lib/components/narrative/metro/metro-error-renderer.tsx
Comment (miles-grant-ibigroup) [reply lib/components/narrative/metro/metro-error-renderer.tsx:1]: It was, it was then merged into this one

File: lib/actions/apiV2.js
Comment (daniel-heppner-ibigroup) [reply lib/actions/apiV2.js:876]: Yep, great catch. That was happening because I added a new type, "SUBMODE" which behaves just like "CHECKBOX", but I forgot to add that to the utility function which converts the true/false string into an actual boolean, so it was staying as a string.

File: lib/components/app/call-taker-panel.js
Comment (daniel-heppner-ibigroup) [reply lib/components/app/call-taker-panel.js:126]: No. We could. Is that needed for call taker?

File: lib/actions/apiV2.js
Comment (daniel-heppner-ibigroup) [reply lib/actions/apiV2.js:876]: Fixed in OTP UI, new alpha package version added to this PR.

File: lib/components/util/styledIcon.tsx
Comment (daniel-heppner-ibigroup) [reply lib/components/util/styledIcon.tsx:59]: There's no trailing whitespace on my local copy, so there's nothing here for me to remove.

File: lib/actions/apiV2.js
Comment (daniel-heppner-ibigroup) [reply lib/actions/apiV2.js:928]: Moving them outside the loop increases the function's complexity too high. Maybe I can simplify it somewhere else. I don't want to make a new dispatched action for this since that would require a ton of argument passing that would be messy.

Comment (daniel-heppner-ibigroup) [reply lib/actions/apiV2.js:928]: I couldn't find a nice way of doing this. Moving these evaluations outside of the loop increases the function's complexity too much. I think this really should just all be refactored in a future PR.

File: lib/components/narrative/narrative-itineraries.js
Comment (daniel-heppner-ibigroup) [reply lib/components/narrative/narrative-itineraries.js:1]: It was a PR into this one along with the other error stuff

File: lib/components/narrative/metro/metro-error-renderer.tsx
Comment (daniel-heppner-ibigroup) [reply lib/components/narrative/metro/metro-error-renderer.tsx:56]: @miles-grant-ibigroup

File: i18n/fr.yml
Comment (daniel-heppner-ibigroup) [reply i18n/fr.yml:749]: Oh shoot! Bad merge

**Comment by miles-grant-ibigroup** [comment]
I've addressed everything assigned to me

File: lib/components/form/batch-settings.tsx
Comment (binh-dam-ibigroup) [inline lib/components/form/batch-settings.tsx:126]: Just thought of this... Changes from #861 need to be merged into this PR.

File: lib/components/form/batch-styled.ts
Comment (binh-dam-ibigroup) [inline lib/components/form/batch-styled.ts:85]: Nice-to-have: The top margin of the `<input>` elements, set by bootstrap, should probably be removed in the popups.

File: lib/components/form/batch-settings.tsx
Comment (binh-dam-ibigroup) [inline lib/components/form/batch-settings.tsx:180]: Remove the "group" role per client request, so that the `<fieldset>` element can be correctly announced by assistive technology.

**Review by binh-dam-ibigroup — CHANGES_REQUESTED** [review]
Need features from #861 and another a11y tweak. Also, the checkboxes still don't have the desired effect.

File: lib/components/form/batch-settings.tsx
Comment (daniel-heppner-ibigroup) [reply lib/components/form/batch-settings.tsx:180]: Right, this had been mentioned to me, thanks for the reminder. I replaced the div with a fragment.

File: lib/components/form/batch-styled.ts
Comment (daniel-heppner-ibigroup) [reply lib/components/form/batch-styled.ts:85]: That makes the alignment much better! Thank you!

**Review by binh-dam-ibigroup — COMMENTED** [review]
Much better now, seeing the light! The basic functionality works now. There is some polishing needed but that can be done in a subsequent PR. I'll approve once the non-alpha version is released.

**Comment by daniel-heppner-ibigroup** [comment]
Current plan is to let @philip-cline review this while we wait for all the alpha builds to be merged.

File: package.json
Comment (binh-dam-ibigroup) [inline package.json]: Sort dependencies. Is this the correct version number?

File: lib/actions/api.js
Comment (binh-dam-ibigroup) [inline lib/actions/api.js:1]: Could you address the "defined but never used" warnings for this file?

**Review by binh-dam-ibigroup — COMMENTED** [review]
One more thing about passing custom mode names.

**Comment by binh-dam-ibigroup** [comment]
@daniel-heppner-ibigroup Please do the cleanup and advise regarding versions and passing custom mode names.

File: lib/components/narrative/default/itinerary-description.tsx
Comment (miles-grant-ibigroup) [reply lib/components/narrative/default/itinerary-description.tsx]: Addressed in 0382c541cb579e3cf41f2545bc060a7e1aa85a9e

File: package.json
Comment (binh-dam-ibigroup) [reply package.json]: This PR should use `@opentripplanner/trip-form@^3.1.0`.

File: lib/components/form/batch-settings.tsx
Comment (daniel-heppner-ibigroup) [reply lib/components/form/batch-settings.tsx:115]: I think I had in my head that is handled by i18n stuff, which I realize is not right. I can add another step to this to pull in those strings.

**Comment by daniel-heppner-ibigroup** [comment]
I added functionality for custom mode names using the old `getFormattedMode` function. However, this was always possible thanks to the i18n keys.

**Comment by binh-dam-ibigroup** [comment]
🚨 @daniel-heppner-ibigroup  and @miles-grant-ibigroup The type check is failing.

File: lib/actions/apiV2.js
Comment (binh-dam-ibigroup) [inline lib/actions/apiV2.js:7]: Use `const { randId } = `coreUtils.storage`.

File: lib/actions/apiV2.js
Comment (binh-dam-ibigroup) [inline lib/actions/apiV2.js:784]: Add JSDoc

File: lib/actions/ui.js
Comment (binh-dam-ibigroup) [inline lib/actions/ui.js:311]: Also delete the "remove" phrase from the comment above.

File: lib/components/form/batch-settings.tsx
Comment (binh-dam-ibigroup) [inline lib/components/form/batch-settings.tsx:182]: Deduplicate comment.

File: lib/components/form/batch-settings.tsx
Comment (binh-dam-ibigroup) [inline lib/components/form/batch-settings.tsx:202]: Use `intl.formatMessage` here.

File: lib/components/form/batch-styled.ts
Comment (binh-dam-ibigroup) [inline lib/components/form/batch-styled.ts:49]: Sort CSS.

File: lib/components/narrative/narrative-itineraries.js
Comment (binh-dam-ibigroup) [inline lib/components/narrative/narrative-itineraries.js:487]: Remove or update this comment.

File: lib/components/narrative/metro/metro-error-renderer.tsx
Comment (binh-dam-ibigroup) [inline lib/components/narrative/metro/metro-error-renderer.tsx:65]: Note for another PR: After pasting a URL in the browser's search bar, the error message might appear, but can remain even if results were updated.

image

File: lib/components/form/call-taker/mode-dropdown.tsx
Comment (binh-dam-ibigroup) [inline lib/components/form/call-taker/mode-dropdown.tsx:40]: Note for another PR: not sure if it is a config issue, but the "Rail only" option seems to be too strict, returning no itineraries whereas the "transit" option produces a Walk-MAX itinerary.

**Review by binh-dam-ibigroup — APPROVED** [review]
Nice work (that took too long)! Please see the minor tweaks suggested and fix the TypeScript errors. Include #921 to fix the broken date/time selector styles.

File: lib/components/form/call-taker/mode-dropdown.tsx
Comment (daniel-heppner-ibigroup) [reply lib/components/form/call-taker/mode-dropdown.tsx:40]: Probably just bad config.

File: lib/components/form/call-taker/advanced-options.js
Comment (daniel-heppner-ibigroup) [reply lib/components/form/call-taker/advanced-options.js]: This is my first time seeing this... this code is pretty wild. Hopefully we can move calltaker out of OTP-RR soon.

**Review by philip-cline — APPROVED** [review]
I tried to test this pretty rigorously in the following areas and couldn't find anything wrong: 
- call-taker
- trip saving
- slow network speeds
- various different configurations

So it's looking pretty good for me. Fixed that one whitespace issue.

Great work !

**Pull request merged**
